Shopify/cli
# Improve global cli detection

### WHY are these changes introduced?
Our way to detect if we were running globally or locally wasn't very robust and failed in some cases.



### WHAT is this pull request doing?
- Implement a new way to decide if we are running globally or not by using `npm prefix`.
  - `npm prefix` tells you the path to the root of your project, and with `argv` we know what javascript file we are executing. If `argv` starts with same path as `npm prefix` we are certain to be running a local javascript file (inside node_modules) 

- Move `sniffForPath` to a more appropriate location.
- Add `sniffForJson` to detect if the `--json` flag is being used.
- Don't show the global CLI warning if using `--json`
- Force hydrogen's `cli-kit` and `plugin-cloudflare` resolution to use the local version of those packages.


### How to test your changes?
This branch has been released to experimental: `0.0.0-experimental-20240610101312)`

You can test basic usage with an app, but basically changes in this PR are covered by tests in `is-global`.
Running `npm run shopify` inside an app project (that doesn't have a local CLI dependency) should run the global CLI and detect it has such. (previously it was detected as a local CLI)

For instance, running `app info` will show a message like `To change these, run "npm run dev -- --reset"`. That message should change if we detect a local vs a global CLI



### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
